Fix issues #29 and #35 - Modeling Assistants for Lifelines, Messages

cdamus opened the pull request:
## Overview

This pull request comprises several commits, some laying down pre-requisites for the main implementation of the assistants feature.  These include bug fixes, enhancements, and extensions of Papyrus frameworks tailored for this project's architecture in which so much of the behaviour that in GMF/Papyrus is pluggable on extension points is instead delegated to the *Logical Model API*.

## Logical model Bug fixes

Commit bfe106c8e01866f7d3188bdef7d7e8fb48152118 fixes problems in the *Logical Model API* implementation of message creation commands, including:

- the commands always attached both ends of the message `Connector` view to the lifeline head views.  Now the attachment is to the lifeline body.  Note that this does not (yet) support `create` messages, `delete` messages, nor attachment to execution specifications
- the preparation of commands when mousing over the diagram would create reams and reams  of message `Connector` views attached to the diagram that were not deleted but ended up as dangling HREFs.  This was a failure to account for the fact that the `source` and `target` references of the `Edge` class have opposites in the `View` class, so they have to be manipulated using undoable commands instead of directly on the not-yet-attached new instance

## Diagram Filters

Commit 78f742d2fd61b2714799fd907d38023cfd33afde defines filters for constraining the applicability of assistants in the diagrams, with conditions on `View`s and the `Diagram`s containing them.  The purpose of this is to allow the assistants model for the *Lightweight Sequence Diagram* to be restricted only to that diagram type.  It seems like these filters could be generally useful for the various Papyrus models that employ the `Filter` model, so I use a namespace appropriate to that general usage.  But **perhaps we should nonetheless use the `org.eclipse.papyrus.uml.sequence` namespace** at least for the time being?

## Custom Modeling Assistant Provider

Commit 57d98c92c0be0e61459c4e51600712f7e4c5a250 addresses a pseudo-limitation of the Papyrus modelled assistant framework, which uses the `ViewService` to filter out of the connection handle pop-up menu and relation types for which the diagram cannot create a view.  This is entirely reasonable in a GMF/Papyrus context (hence "pseudo" limitation) because it is a peculiarity of the *Lightweight Sequence Diagram* architecture that so much editing behaviour is delegated from the diagram framework to the *Logical Model API*.

## Modeling Assistants Definition

The fourth commit, 48e441f992d5e9bf080991ae48e5892dd999f911, actually defines assistants for lifelines, both kinds of execution specification, and all of the message types enumerated by Papyrus.  This requires new edit policies, including specializations of the Papyrus implementations of the diagram assistant edit policies, including:

- popup bar for sequence diagram, activated by mouse click in the interaction frame for creation of lifelines (not just mouse hover)
- connection handles for sequence diagram
- graphical node policy for creation of message connections

To get the assistants to show up requires implementation of the Papyrus preferences for the *Lightweight Sequence Diagram*, which is done herein.  Also more additions to the *Logical Model API*, especially for creation of execution specifications which previously required having the execution's signature on hand, which isn't practical for the tool based creation scenario.

Finally, with the addition of more edit policies that override Papyrus's default implementations, the `NoPapyrusEditPolicyProvider` work-around expanded further, so I refactored it to make that (I hope) a bit more manageable.

## Outstanding Problems

A major outstanding problem in this pull request is bad aim in the placement of messages and execution specifications on lifeline bodies, whether via the palette tool or the diagram assistant.  Any suggestions on what kind of coördinate transformation or other computation is missing will be most welcome.

Some unit tests are updated for changes or added for new code, but there is no real coverage of the tool palette nor especially the assistants.  Some JUnit tests for the creation tools will follow some infrastructural (test framework) work done in the scope of pull #290.  Test coverage of the assistants could be done with JUnit but not actually invoking them in the diagram:  it would only be necessary to test that the assistant service provides the correct element types for any scenario; automation of the creation tool itself can test the diagram edit parts independently of the assistants (which framework is tested by Papyrus).

planger commented:
While testing I also got an error dialog popping up:
![image](https://user-images.githubusercontent.com/588090/40106345-db56c642-58f5-11e8-93f9-ecb7836cd39a.png)

```
!ENTRY org.eclipse.ui 4 4 2018-05-16 10:41:40.811
!MESSAGE An internal error has occurred.
!STACK 0
java.lang.IllegalArgumentException: dependency cycle: [Lifeline{G}(AnchorsModel::LifelineBodyAnchors::LeftLine, Shape_Lifeline_Header), MessageOccurrenceSpecification(AnchorsModel::LifelineBodyAnchors::left-send), Message(AnchorsModel::LifelineBodyAnchors::leftToRight, Edge_Message), MessageOccurrenceSpecification(AnchorsModel::LifelineBodyAnchors::right-recv), MessageOccurrenceSpecification(AnchorsModel::LifelineBodyAnchors::right-send), MessageOccurrenceSpecification(AnchorsModel::LifelineBodyAnchors::<unnamed>), Message(AnchorsModel::LifelineBodyAnchors::<unnamed>, Edge_Message)]
	at org.eclipse.papyrus.uml.interaction.internal.graph.EdgeImpl.checkForCycle(EdgeImpl.java:156)
	at org.eclipse.papyrus.uml.interaction.internal.graph.EdgeImpl.<init>(EdgeImpl.java:41)
	at org.eclipse.papyrus.uml.interaction.internal.graph.GraphComputer.edge(GraphComputer.java:94)
	at org.eclipse.papyrus.uml.interaction.internal.graph.GraphComputer.edge(GraphComputer.java:90)
	at org.eclipse.papyrus.uml.interaction.internal.graph.GraphComputer.lambda$0(GraphComputer.java:108)
	at java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:193)
	at java.util.stream.DistinctOps$1$2.accept(DistinctOps.java:175)
	at java.util.Iterator.forEachRemaining(Iterator.java:116)
	at java.util.Spliterators$IteratorSpliterator.forEachRemaining(Spliterators.java:1801)
	at java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:481)
	at java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:471)
	at java.util.stream.ReduceOps$ReduceOp.evaluateSequential(ReduceOps.java:708)
	at java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
	at java.util.stream.ReferencePipeline.collect(ReferencePipeline.java:499)
	at org.eclipse.papyrus.uml.interaction.internal.graph.GraphComputer.edge(GraphComputer.java:108)
	at org.eclipse.papyrus.uml.interaction.internal.graph.GraphComputer.edge(GraphComputer.java:120)
	at org.eclipse.papyrus.uml.interaction.internal.graph.GraphComputer$EdgeFactory.caseMessage(GraphComputer.java:155)
	at org.eclipse.papyrus.uml.interaction.internal.graph.GraphComputer$EdgeFactory.caseMessage(GraphComputer.java:1)
	at org.eclipse.uml2.uml.util.UMLSwitch.doSwitch(UMLSwitch.java:4667)
	at org.eclipse.emf.ecore.util.Switch.doSwitch(Switch.java:53)
	at org.eclipse.emf.ecore.util.Switch.doSwitch(Switch.java:69)
	at org.eclipse.papyrus.uml.interaction.internal.graph.GraphComputer.compute(GraphComputer.java:85)
	at org.eclipse.papyrus.uml.interaction.internal.graph.GraphImpl.<init>(GraphImpl.java:77)
	at org.eclipse.papyrus.uml.interaction.internal.graph.GraphImpl.getInstance(GraphImpl.java:96)
	at org.eclipse.papyrus.uml.interaction.graph.Graph.compute(Graph.java:179)
	at org.eclipse.papyrus.uml.interaction.internal.model.impl.InteractionModelBuilder.<init>(InteractionModelBuilder.java:44)
	at org.eclipse.papyrus.uml.interaction.internal.model.impl.InteractionModelBuilder.getInstance(InteractionModelBuilder.java:57)
	at org.eclipse.papyrus.uml.interaction.model.MInteraction.getInstance(MInteraction.java:56)
	at org.eclipse.papyrus.uml.diagram.sequence.runtime.internal.edit.policies.LogicalModelCreationEditPolicy.getCreateElementAndViewCommand(LogicalModelCreationEditPolicy.java:54)
	at org.eclipse.gmf.runtime.diagram.ui.editpolicies.CreationEditPolicy.getCommand(CreationEditPolicy.java:87)
	at org.eclipse.papyrus.uml.diagram.sequence.runtime.internal.edit.policies.InteractionCreationEditPolicy.getCommand(InteractionCreationEditPolicy.java:41)
	at org.eclipse.gef.editparts.AbstractEditPart.getCommand(AbstractEditPart.java:502)
	at org.eclipse.gmf.runtime.diagram.ui.editparts.GraphicalEditPart.access$1(GraphicalEditPart.java:1)
	at org.eclipse.gmf.runtime.diagram.ui.editparts.GraphicalEditPart$1.run(GraphicalEditPart.java:482)
	at org.eclipse.papyrus.infra.emf.readonly.PapyrusROTransactionalEditingDomain.runExclusive(PapyrusROTransactionalEditingDomain.java:271)
	at org.eclipse.gmf.runtime.diagram.ui.editparts.GraphicalEditPart.getCommand(GraphicalEditPart.java:477)
	at org.eclipse.papyrus.uml.diagram.sequence.runtime.internal.edit.policies.SequenceDiagramPopupBarEditPolicy$DefaultTool.getCommand(SequenceDiagramPopupBarEditPolicy.java:167)
	at org.eclipse.gmf.runtime.diagram.ui.tools.AbstractPopupBarTool.isCommandEnabled(AbstractPopupBarTool.java:96)
	at org.eclipse.papyrus.infra.gmfdiag.common.editpolicies.PapyrusPopupBarEditPolicy.addPopupBarDescriptor(PapyrusPopupBarEditPolicy.java:325)
	at org.eclipse.papyrus.infra.gmfdiag.common.editpolicies.PapyrusPopupBarEditPolicy.addPopupBarDescriptor(PapyrusPopupBarEditPolicy.java:347)
	at org.eclipse.papyrus.infra.gmfdiag.common.editpolicies.PapyrusPopupBarEditPolicy.addPopupBarDescriptor(PapyrusPopupBarEditPolicy.java:360)
	at org.eclipse.papyrus.infra.gmfdiag.common.editpolicies.PapyrusPopupBarEditPolicy.populatePopupBars(PapyrusPopupBarEditPolicy.java:198)
	at org.eclipse.papyrus.infra.gmfdiag.common.editpolicies.PapyrusPopupBarEditPolicy.showDiagramAssistant(PapyrusPopupBarEditPolicy.java:509)
	at org.eclipse.papyrus.uml.diagram.sequence.runtime.internal.edit.policies.SequenceDiagramPopupBarEditPolicy.showDiagramAssistant(SequenceDiagramPopupBarEditPolicy.java:53)
	at org.eclipse.gmf.runtime.diagram.ui.editpolicies.DiagramAssistantEditPolicy$ShowDiagramAssistantJob.runInUIThread(DiagramAssistantEditPolicy.java:98)
	at org.eclipse.ui.progress.UIJob.lambda$0(UIJob.java:95)
	at org.eclipse.swt.widgets.RunnableLock.run(RunnableLock.java:37)
	at org.eclipse.swt.widgets.Synchronizer.runAsyncMessages(Synchronizer.java:182)
	at org.eclipse.swt.widgets.Display.runAsyncMessages(Display.java:4577)
	at org.eclipse.swt.widgets.Display.readAndDispatch(Display.java:4186)
	at org.eclipse.e4.ui.internal.workbench.swt.PartRenderingEngine$5.run(PartRenderingEngine.java:1150)
	at org.eclipse.core.databinding.observable.Realm.runWithDefault(Realm.java:336)
	at org.eclipse.e4.ui.internal.workbench.swt.PartRenderingEngine.run(PartRenderingEngine.java:1039)
	at org.eclipse.e4.ui.internal.workbench.E4Workbench.createAndRunUI(E4Workbench.java:153)
	at org.eclipse.ui.internal.Workbench.lambda$3(Workbench.java:680)
	at org.eclipse.core.databinding.observable.Realm.runWithDefault(Realm.java:336)
	at org.eclipse.ui.internal.Workbench.createAndRunWorkbench(Workbench.java:594)
	at org.eclipse.ui.PlatformUI.createAndRunWorkbench(PlatformUI.java:148)
	at org.eclipse.ui.internal.ide.application.IDEApplication.start(IDEApplication.java:151)
	at org.eclipse.equinox.internal.app.EclipseAppHandle.run(EclipseAppHandle.java:196)
	at org.eclipse.core.runtime.internal.adaptor.EclipseAppLauncher.runApplication(EclipseAppLauncher.java:134)
	at org.eclipse.core.runtime.internal.adaptor.EclipseAppLauncher.start(EclipseAppLauncher.java:104)
	at org.eclipse.core.runtime.adaptor.EclipseStarter.run(EclipseStarter.java:388)
	at org.eclipse.core.runtime.adaptor.EclipseStarter.run(EclipseStarter.java:243)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.eclipse.equinox.launcher.Main.invokeFramework(Main.java:653)
	at org.eclipse.equinox.launcher.Main.basicRun(Main.java:590)
	at org.eclipse.equinox.launcher.Main.run(Main.java:1499)
	at org.eclipse.equinox.launcher.Main.main(Main.java:1472)
```

cdamus commented:
I had originally customized the location of the pop-up bar to be right at the mouse precisely so that moving to click it would not involve travel over the interaction compartment, which causes the pop-up bar to disappear.  But then when I added the connection handles, they were partially occluded by the pop-up bar, so I moved it away again.  Oops.  I don't want to require an actual mouse click to dismiss the pop-up bar, so probably some kind of delay is in order.  I'll figure something out.

The other problem of error dialogs indicates something broken in the interaction semantics, perhaps in the ordering of fragments.  It's not related to the diagram assistants as such.  The graph API is strict about dependency cycles; probably the graph builder needs to be more tolerant of invalid interactions and generate graphs without cycles, even if it means missing ordering dependencies that are manifest in the model.
